libretroshare/src/util/i2pcommon.h

```diff
@@ -0,0 +1,224 @@
+	ECDSA_SHA512_P521 = 3,
+	RSA_SHA256_2048   = 4,
+	RSA_SHA384_3072   = 5,
+	RSA_SHA512_4096   = 6,
+	EdDSA_SHA512_Ed25519   = 7,
+	EdDSA_SHA512_Ed25519ph = 8,
+	RedDSA_SHA512_Ed25519  = 11
+};
+
+/*
+ * public
+	Type		Type Code	Total Public Key Length	Usage
+	ElGamal		0			256						All Router Identities and Destinations
+	P256		1			64						Reserved, see proposal 145
+	P384		2			96						Reserved, see proposal 145
+	P521		3			132						Reserved, see proposal 145
+	X25519		4			32						Not for use in key certs. See proposal 144
+	reserved	65280-65534	 						Reserved for experimental use
+	reserved	65535								Reserved for future expansion
+
+ * private
+	Type	Length (bytes)	Since	Usage
+	ElGamal	256						All Router Identities and Destinations
+	P256	32				TBD		Reserved, see proposal 145
+	P384	48				TBD		Reserved, see proposal 145
+	P521	66				TBD		Reserved, see proposal 145
+	X25519	32				0.9.38	Little-endian. See proposal 144
+*/
+enum class CryptoKeyType : uint16_t {
+	ElGamal = 0,
+	P256    = 1,
+	P384    = 2,
+	P521    = 3,
+	X25519  = 4
+};
+
+static const std::array<std::pair<uint16_t, uint16_t>, 5> cryptoKeyLengths {
+	/*CryptoKeyType::ElGamal*/ std::make_pair<uint16_t, uint16_t>(256, 256),
+	/*CryptoKeyType::P256,  */ std::make_pair<uint16_t, uint16_t>( 64,  32),
+	/*CryptoKeyType::P384,  */ std::make_pair<uint16_t, uint16_t>( 96,  48),
+	/*CryptoKeyType::P521,  */ std::make_pair<uint16_t, uint16_t>(132,  66),
+	/*CryptoKeyType::X25519,*/ std::make_pair<uint16_t, uint16_t>( 32,  32),
+};
+
+static const std::array<std::pair<uint16_t, uint16_t>, 12> signingKeyLengths {
+	/*SigningKeyType::DSA_SHA1,              */ std::make_pair<uint16_t, uint16_t>(128, 128),
+	/*SigningKeyType::ECDSA_SHA256_P256,     */ std::make_pair<uint16_t, uint16_t>( 64,  32),
+	/*SigningKeyType::ECDSA_SHA384_P384,     */ std::make_pair<uint16_t, uint16_t>( 96,  48),
+	/*SigningKeyType::ECDSA_SHA512_P521,     */ std::make_pair<uint16_t, uint16_t>(132,  66),
+	/*SigningKeyType::RSA_SHA256_2048,       */ std::make_pair<uint16_t, uint16_t>(256, 512),
+	/*SigningKeyType::RSA_SHA384_3072,       */ std::make_pair<uint16_t, uint16_t>(384, 768),
+	/*SigningKeyType::RSA_SHA512_4096,       */ std::make_pair<uint16_t, uint16_t>(512,1024),
+	/*SigningKeyType::EdDSA_SHA512_Ed25519   */ std::make_pair<uint16_t, uint16_t>( 32,  32),
+	/*SigningKeyType::EdDSA_SHA512_Ed25519ph */ std::make_pair<uint16_t, uint16_t>( 32,  32),
+	/*reserved (GOST)                        */ std::make_pair<uint16_t, uint16_t>( 64,   0),
+	/*reserved (GOST)                        */ std::make_pair<uint16_t, uint16_t>(128,   0),
+	/*SigningKeyType::RedDSA_SHA512_Ed25519  */ std::make_pair<uint16_t, uint16_t>( 32,  32),
+};
+
+
+/**
+ * @brief generateNameSuffix Generates a base32 name suffix for tunnel identification
+ * @param len lenght of random bytes, will be expanded by base32 encoding
+ * @return base32 string
+ *
+ * RSRandom::random_alphaNumericString can return very weird looking strings like: ,,@z+M
```

Comment:
That seems a bug or at least a misleading function name in  RsRandom::random_alphaNumericString

Reply:
Indeed, thats the line that generates the characters:
`s += (char)( (random_u32()%94) + 33) ;`